src/s3_plugin.js

```diff
@@ -118,6 +129,23 @@ module.exports = class S3Plugin {
     })
   }
 
+  addGitHashToBasePath(basePath) {
+    return new Promise((resolve, reject) => {
+      if (!this.options.git || !this.options.git.addGitHash) resolve({basePath})
+      var DirectoryLevelToInsertHashConf = this.options.git.DirectoryLevelToInsertHash
+      gitsha(function(err, sha) {
+        if (err) reject(err)
+        var basePathParts = basePath.split(S3_PATH_SEP)
+        var DirectoryLevelToInsertHash = basePathParts.length - 2 // set directory level to insert git SHA; default is the last\deepest directory
+        if (DirectoryLevelToInsertHashConf) { // unless DirectoryLevelToInsertHashConf is defined
+          DirectoryLevelToInsertHash = DirectoryLevelToInsertHashConf - 1 // use the defined directory level and convert it to array index
+        }
+        basePathParts[DirectoryLevelToInsertHash] += `.${sha.substring(7, 0)}`
+        resolve(basePathParts.join(S3_PATH_SEP))
```

Comment:
Before your resolve why not set `options.basePath` that way you can keep the promise chain flat as per my comment above